### ISSUE TITLE
Feature Request: customizable early_stopping_tolerance

### ISSUE THREAD

**Issue by kryptonite0** [microsoft/LightGBM#2526]
I have a situation where the default numerical tolerance (0.001) for early stopping is too large. My target has a gamma distribution and the LGB Regressor reaches convergence too early, when the numerous low target values are well approximated by the model, but the few large values are still underestimated. When I deactivate early stopping, I can see the loss metric still improving at the 4th or more decimal digit, past the best iteration reached during early stopping. It would be great to be able to set manually the tolerance.

**Comment by StrikerRUS** [comment]
@kryptonite0 Can you please provide a reproducible example or training logs at least?

As I know, we do not have any **default** numerical tolerance. Refer to:
https://github.com/microsoft/LightGBM/blob/fc991c9d7e07fd8e460ea24d1f5609f61fd2af3f/python-package/lightgbm/callback.py#L227
https://docs.python.org/3/library/operator.html#operator.lt

Linking https://github.com/dmlc/xgboost/issues/4982 here.

**Comment by StrikerRUS** [comment]
Ping @kryptonite0 

As you can see from examples, there is no "default numerical tolerance (0.001)":
https://github.com/microsoft/LightGBM/blob/785e477856b9e9e9e69123caf175dbb8455c3b11/examples/python-guide/advanced_example.py#L52-L59

```
[1]	training's binary_logloss: 0.680894
[2]	training's binary_logloss: 0.672151
[3]	training's binary_logloss: 0.664753
[4]	training's binary_logloss: 0.656185
[5]	training's binary_logloss: 0.648174
[6]	training's binary_logloss: 0.641671
[7]	training's binary_logloss: 0.635597
[8]	training's binary_logloss: 0.628874
[9]	training's binary_logloss: 0.622432
[10]	training's binary_logloss: 0.616403
Finished first 10 rounds...
```

**Comment by StrikerRUS** [comment]
At present we do not have any "default numerical tolerance". But having customizable early stopping tolerance might be useful in some cases.

**Comment by StrikerRUS** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.

**Comment by StrikerRUS** [comment]
#4580 implemented this feature request for Python-package. Thank you very much @jmoralez !
I think we should reuse this issue as a feature request for adding the same functionality into R-package and core cpp code to not split the discussion. Refer to https://github.com/microsoft/LightGBM/pull/4580#issuecomment-910714636. So, I'm not excluding this issue from #2302 for now.
